services/galley/src/Galley/API/Update.hs

```diff
@@ -105,13 +113,114 @@ unblockConv (usr ::: conn ::: cnv) = do
     conv' <- acceptOne2One usr conv conn
     setStatus status200 . json <$> conversationView usr conv'
 
-joinConversation :: UserId ::: ConnId ::: ConvId ::: JSON -> Galley Response
-joinConversation (zusr ::: zcon ::: cnv ::: _) = do
+updateConversationAccess :: UserId ::: ConnId ::: ConvId ::: Request ::: JSON -> Galley Response
+updateConversationAccess (usr ::: zcon ::: cnv ::: req ::: _ ) = do
+    body <- fromBody req invalidPayload
+    let targetAccess = Set.fromList (toList (cupAccess body))
+    -- checks and balances
+    when (PrivateAccess `elem` targetAccess) $
+        throwM invalidTargetAccess
+    (bots, users) <- botsAndUsers <$> Data.members cnv
+    unless (usr `isMember` users) $
+        throwM convNotFound
+    conv <- Data.conversation cnv >>= ifNothing convNotFound
+    let currentAccess = Set.fromList (toList $ Data.convAccess conv)
+    case Data.convType conv of
+        SelfConv    -> throwM invalidSelfOp
+        One2OneConv -> throwM invalidOne2OneOp
+        ConnectConv -> throwM invalidConnectOp
+        _           -> return ()
+    -- remove conversation codes if CodeAccess is revoked
+    when (CodeAccess `elem` currentAccess && CodeAccess `notElem` targetAccess) $ do
```

Comment:
Since only team members can change access for team conversations, better if we do the team membership first otherwise we might revoke codes in team conversations